### ISSUE TITLE
[service-bus] Message batching

### ISSUE THREAD

**Issue by richardpark-msft** [Azure/azure-sdk-for-go#15136]
Track 2 added in the ability to create a message batch - basically, it's a "append only" collection of messages which are encoded and sent as a single message. This can be a big efficiency boost for senders as it allows the combining of several small messages into one, eliminating some per-message send overhead which can add up.

Interfaces used for illustration - current policy for the Go SDK (from what I can see) is to use concrete types:

```go
type ServiceBusMessageBatch interface {
  AddMessage(msg *Message) bool
}

type interface ServiceBusSender {
  // other methods elided...

  // looks at the current link for the 'max batch size', creates a message batch
  // NOTE this does actually need to potentially initialize the link so it needs to take a context.
  CreateMessageBatch(ctx context.Context) (*ServiceBusMessageBatch, error)
}

// usage

batch, err := sender.CreateMessageBatch(context.TODO())

// NOTE: in other languages where we throw exceptions this check is less
// complicated. Perhaps we'd like to have a designated error (message too big) 
// so we just have the single return value instead of having `added` and `err`?
if added, err := batch.TryAdd(message); !added {
  // message failed to add - decisions to make:
  if err != nil {
      // If there was an internal encoding failure the best they can do is drop this message.
      return
  }

  // otherwise the only reason we failed here was because the message was too big to fit
  // into the existing batch. So we can:
  // a. Send the batch immediately and then start a new batch with this message
  // b. Discard this message

  // There is also the potential pitfall that the message is too large, even on it's own,
  // to fit into a batch. All of this is determinable using the .TryAdd() call and checking
  // it's return value.
}

sender.SendMessageBatch(context.TODO(), batch)

```

References:
- [batch.go](https://github.com/richardpark-msft/azure-sdk-for-go/blob/e68e55fb55079816240f0dab9991215d7d33885a/sdk/servicebus/batch.go#L106)

**Comment by richardpark-msft** [comment]
There is some code related to batching here: https://github.com/richardpark-msft/azure-sdk-for-go/blob/sb-track-2/sdk/servicebus/batch.go

It looks like it's hard coding the max message size based on the messaging tier, which should not be done (it can be dynamically retrieved from the link).

**Comment by richardpark-msft** [comment]
Even better - the .Add signature in the existing batch code already has the .Add() (bool, error) signature:

https://github.com/richardpark-msft/azure-sdk-for-go/blob/e68e55fb55079816240f0dab9991215d7d33885a/sdk/servicebus/batch.go#L106

**Comment by richardpark-msft** [comment]
Implemented in #15561
